### ISSUE TITLE
Long natural-language prompts intermittently degenerate into token id 0 ("!") — silent KV-cache corruption

### ISSUE THREAD

**Issue by weissmuster** [jundot/omlx#3777]
**Type:** bug · **Component:** KV cache / long-context prefill · **Severity:** high (silent wrong output, no error surfaced)

## Summary

With prompts of roughly 60k tokens and above, generation collapses: the model emits nothing but
the character `!` (token id 0) until `max_tokens` is reached. The very **first** sampled token is
already token 0, so this is not a repetition loop — it is a collapsed logit distribution at the end
of prefill. Nothing is logged as an error; the request returns HTTP 200 with a full-length response.

Enabling TurboQuant KV compression (`turboquant_kv_enabled: true`, 8-bit) makes the same prompts
work reliably, which points at the default KV-cache path as the origin.

The failure affects **both** a hybrid MoE model (Qwen3.6-35B-A3B) and a dense model (Qwen3.8-27B),
so it does not look architecture-specific.

## Environment

| | |
|---|---|
| omlx | 0.6.4 (Homebrew `jundot/omlx/omlx`, `brew services`) |
| mlx / mlx-lm / mlx-vlm | 0.32.0 / 0.31.3 / 0.6.3 |
| dflash-mlx | 0.1.10+omlx.7 |
| transformers | 5.12.1 |
| Hardware | Mac Studio, Apple M1 Max, 32 GB unified memory |
| OS | macOS 26.5.1 (25F80) |
| `iogpu.wired_limit_mb` | 27648 |
| Models | `lmstudio-community/Qwen3.6-35B-A3B-MLX-4bit`, `lmstudio-community/Qwen3.8-27B-MLX-4bit` |
| Endpoint used | Anthropic-compatible `/v1/messages` (same behaviour is expected on `/v1/chat/completions`) |

## Reproduction

`repro_omlx.py` (attached) needs only the standard library. It downloads a public-domain text
(Project Gutenberg, *Moby-Dick*), builds prompts of increasing length and reports whether the
response consists solely of `!`:

```bash
python3 repro_omlx.py "<model-id>" "http://localhost:8000" "<api-key>"
```

Two details matter:

1. **Use real prose.** Prompts assembled from random digits did *not* trigger the bug in our
   tests, even at 240k tokens. Natural-language text does.
2. **Defeat the prefix cache.** The per-run marker must go at the **start** of the prompt.
   With the marker at the end, the prefix cache serves a stored state and you measure that
   instead of a fresh prefill (see "Cache poisoning" below).

### Observed with the public-domain corpus (TurboQuant off)

| input tokens | run 1 | run 2 | prefill |
|---|---|---|---|
| 51,583 | ok | ok | 144–154 s |
| 61,153 | ok | **BROKEN** | 178–179 s |
| 67,602 | **BROKEN** | ok | 188–190 s |

**The failure is probabilistic**, not a hard threshold: the same prompt succeeds and fails across
runs. With a different corpus (1.1 MB of German/English technical prose) it was deterministic at
higher lengths — 67,980 tokens clean, 78,468 and 85,571 tokens broken in 2 of 2 runs each.

### Effect of TurboQuant KV compression (same prompts, same session)

| model | input tokens | default KV path | `turboquant_kv_enabled: true` (8-bit) |
|---|---|---|---|
| Qwen3.6-35B-A3B (hybrid MoE) | 78,468 | BROKEN (2/2) | ok (2/2) |
| Qwen3.6-35B-A3B | 85,571 / 85,713 | BROKEN (2/2) | ok, incl. a 289 s cold prefill |
| Qwen3.6-35B-A3B | 104,267 | — | ok (2/2) |
| Qwen3.6-35B-A3B | 123,561 | — | ok (2/2) |
| Qwen3.8-27B (dense) | 78,575 / 85,703 | BROKEN (2/2) | ok, 1,400 s cold prefill |

## What we ruled out

- **Sampling.** The corruption is in the *first* sampled token, so penalties cannot be responsible.
  Tested with `temperature` 1.0/0.7, `top_p` 0.95/0.8, `presence_penalty` 0.0/1.0 — no effect.
  (Note: NaN or ±inf in the logits yields index 0 with this sampler configuration.)
- **Prefix / boundary cache as the *cause*.** The bug reproduces on a cold request with
  `cache_read = 0`; conversely a 75k-token prompt served from a full cache hit came back clean.
- **Memory pressure.** No "Insufficient Memory" entries around the failures. OOM surfaces as
  HTTP 500 or an empty body, not as `!`.
- **Thinking / chat template.** Reproduces with no `thinking` field in the request and with
  `enable_thinking: false` pinned per model.
- **`max_context_window`.** Unset during the failing runs; exceeding it produces a clean HTTP 400.

## Cache poisoning (secondary finding)

A corrupted result gets **stored and replayed**. Re-running an identical failing prompt returned
`!` again in 49 s instead of 1,307 s — a prefix-cache hit on the broken state. This nearly produced
a false conclusion during our own investigation ("the workaround does not help the dense model").
After an incident, `POST /admin/api/ssd-cache/clear` is required to get clean behaviour back.

## No NaN/inf guard in the sampling path

`grep -rl` over `omlx/` and `dflash_mlx/` returns **zero** matches for `isnan`, `isinf` (the seven
hits are incidental words inside `bench_corpora/` and `eval/data/`) and `nan_to_num`. A collapsed or
non-finite logit vector therefore cannot be detected — it silently becomes token id 0. Even without
fixing the root cause, a guard would turn a silent wrong answer into a diagnosable error.

## Suggested next steps

1. **Detect it:** assert finiteness of the logits before sampling (`mx.isnan(logits).any()` /
   `mx.isinf(...)`), log the request id, prompt length and layer, and fail the request instead of
   emitting token 0.
2. **Localise it:** the divergence between the default and the TurboQuant path narrows the search.
   `turboquant_skip_last` is documented as *"Skip last KVCache layer to prevent corruption"*, which
   suggests related failure modes are already known.
3. **Consider** whether chunked prefill boundaries are involved — failing runs logged
   `Skipping cache store … reason=boundary_snapshot_unavailable` and
   `Using boundary cache snapshot … skipping trailing partial block`.

## Workaround for other users

Set per model: `turboquant_kv_enabled: true`, `turboquant_kv_bits: 8`, `turboquant_skip_last: true`.
Additionally set `max_context_window` to a tested value so oversized prompts are rejected with
HTTP 400 rather than silently corrupted. Clear the SSD prompt cache after any incident.

<details>
<summary><b>repro_omlx.py</b> (stdlib only)</summary>

```python
#!/usr/bin/env python3
"""Minimal reproducer: long natural-language prompts degenerate into token id 0 ('!').

Usage:  python3 repro_omlx.py <model-id> [host] [api-key]
Needs no dependencies beyond the standard library. Downloads a public-domain text
(Project Gutenberg) so the prompt is real prose, not synthetic filler — that distinction
matters: prompts built from random digits did NOT trigger the bug in our tests.
"""
import json, sys, time, urllib.request, urllib.error

MODEL = sys.argv[1] if len(sys.argv) > 1 else "mlx-community/Qwen3.6-35B-A3B-4bit"
HOST  = sys.argv[2] if len(sys.argv) > 2 else "http://localhost:8000"
KEY   = sys.argv[3] if len(sys.argv) > 3 else ""
TEXT_URL = "https://www.gutenberg.org/files/2701/2701-0.txt"   # Moby Dick, public domain

print(f"downloading corpus from {TEXT_URL} ...", flush=True)
with urllib.request.urlopen(TEXT_URL, timeout=120) as r:
    CORPUS = r.read().decode("utf-8", "replace")
print(f"corpus: {len(CORPUS):,} chars\n")

def run(target_tokens, attempt):
    body_text = (CORPUS * 3)[: int(target_tokens * 3.6)]          # ~3.6 chars/token for prose
    marker = f"[run {attempt}/{target_tokens}] "                   # marker FIRST -> defeats prefix cache
    payload = json.dumps({"model": MODEL, "max_tokens": 24, "messages": [
        {"role": "user", "content": marker + "Summarise the following text in one sentence.\n\n" + body_text}]}).encode()
    req = urllib.request.Request(f"{HOST}/v1/messages", data=payload, method="POST")
    req.add_header("content-type", "application/json")
    req.add_header("anthropic-version", "2023-06-01")
    if KEY: req.add_header("x-api-key", KEY)
    t0 = time.time()
    try:
        with urllib.request.urlopen(req, timeout=3600) as r:
            d = json.loads(r.read())
    except urllib.error.HTTPError as e:
        return f"HTTP {e.code}", round(time.time()-t0), None, e.read().decode()[:100]
    txt = "".join(c.get("text", "") for c in d.get("content", []) if c.get("type") == "text").strip()
    broken = bool(txt) and set(txt) <= set("!\n ")
    return ("BROKEN" if broken else "ok"), round(time.time()-t0), d.get("usage", {}).get("input_tokens"), txt[:70]

print(f"{'target':>8} {'run':>4} {'status':>7} {'secs':>6} {'input_tokens':>13}  output")
for target in (55_000, 65_000, 72_000):
    for attempt in (1, 2):
        st, secs, inp, out = run(target, attempt)
        print(f"{target:>8,} {attempt:>4} {st:>7} {secs:>6} {str(inp):>13}  {out}", flush=True)
```

</details>

**Comment by jundot** [comment]
@weissmuster I ran your unchanged script six times, for 36 requests total at the reported prompt lengths. All returned normal text, with no `!` collapse. This included 18 cold-cache requests, 12 repeated requests, and 6 SSD-cache restores after a server restart. My setup is an M3 Ultra with a text-only copy of the checkpoint, so this does not rule out an issue specific to your setup.

Without `cache_control`, oMLX 0.6.4 reports `cache_read_input_tokens: 0` even when its automatic prefix cache hits. Please set the log level to **Debug** in **Global Settings**, then retry with a separate empty cache directory. If the issue reproduces, please attach the debug log covering startup through the failed request, along with redacted model settings. I need those to distinguish a fresh prefill failure from replaying previously stored state.

**Comment by weissmuster** [comment]
Thanks for running it that thoroughly, that narrows things down considerably.

I think the difference is the checkpoint. You used a **text-only copy**; mine is the unmodified `lmstudio-community/Qwen3.6-35B-A3B-MLX-4bit`, which is **multimodal**:

```json
"architectures": ["Qwen3_5MoeForConditionalGeneration"],
"vision_config": { ... },
"image_token_id": ...
```

The model directory also carries `preprocessor_config.json`, `processor_config.json` and `video_preprocessor_config.json`.

As a result every request here goes through the VLM path, even though the repro script only ever sends text. My server log has 677 occurrences of:

```
omlx.engine.vlm - INFO - [vlm_stream_generate] ... for request <id>
```

So we may simply be exercising different code. If you still have the unmodified checkpoint, it might be worth one run with that before digging into my logs — that would be cheaper for you than analysing a debug trace.

Second difference: my machine is an **M1 Max / 32 GB**, not an M3 Ultra, so a Metal kernel path specific to the older generation is also possible.

Full setup:

| | |
|---|---|
| Hardware | Apple M1 Max, 32 GB unified memory |
| macOS | 26.5.1 |
| oMLX | 0.6.4 |
| mlx / mlx-lm | 0.32.0 / 0.31.3 |
| Model | `lmstudio-community/Qwen3.6-35B-A3B-MLX-4bit` (unmodified, multimodal) |

I am setting up the run you asked for — Debug log level, a separate empty cache directory, server restarted, then the unchanged script. I will attach the debug log from startup through the first failing request, plus the redacted model settings, as soon as it reproduces.

One more data point that may help: with `turboquant_kv_enabled: true` at 8 bits, I have not seen a single collapse since, including a 100-minute agent session that reached 94k context. Without it, the failures start reliably somewhere above 60k input tokens. That is what made me suspect the standard KV path rather than stored state, but your point about replayed cache state is fair, and the empty cache directory should settle it.

**Comment by weissmuster** [comment]
Follow-up: **I could not reproduce it either.** Two runs, both with `turboquant_kv_enabled: false` and a freshly restarted server, on the unmodified multimodal checkpoint.

**Run 1 — separate empty cache directory, Debug log level:**

```
  target  run  status   secs  input_tokens
  55,000    1      ok    156         51583
  55,000    2      ok    144         51583
  65,000    1      ok    180         61153
  65,000    2      ok    180         61153
  72,000    1      ok    207         67602
  72,000    2      ok    207         67602
```

65k and 72k are exactly where it collapsed reliably before. Verified in the debug log that the setting took effect: `PagedSSDCacheManager updated layer cache signature (40 layers, 2 unique types, turboquant_kv_bits=None, ...)`.

**Run 2 — same script again, now against the cache the first run had populated (17 GB, boundary snapshots at 61k/63k/65k/67k):**

```
  55,000    1      ok      3
  65,000    1      ok      8
  72,000    2      ok      2
```

Two to eight seconds instead of two to three minutes, so the stored snapshots were definitely being replayed — and still no collapse. That seems to rule out replayed cache state as the trigger, at least in this shape.

**What differs from the failing situation:** when I filed the report, the server process had been up for roughly 26 hours and had served a large number of agent sessions, with a cache directory grown to 46 GB. Both runs above used a process that had just started. So my current best guess is accumulated in-process state rather than the standard KV path — which means my original diagnosis in the report is **not supported by the evidence**, and I should not have stated it as confidently as I did.

The `turboquant_kv` workaround did make the failures stop in day-to-day use, but plausibly because it changes the layer cache signature and invalidates everything stored, not because the standard KV path is at fault.

I will keep the debug log level on during normal work and report back with the full trace if it happens again under load. If it would help, I can also leave a long-running server under agent workload and check periodically — that seems closer to the conditions than a short scripted run. Sorry for sending you down a path that my own retest does not confirm.
